Change time series key encoding to use sext format encoded tuples.

jonmeredith opened the pull request:
The previous encoding from leveldb_ts:encode_key stored the length first as a variable int, then the binary which meant that all 1-byte strings appeared before all 2-byte strings ( "A", "B", "AA", "AB", "BA" ) rather than in lexicographical order ("A", "AA", "AB", "B", "BA") which could cause future problems for ranges of family keys.

Instead of using the custom encoding in eleveldb_ts:encode_keys, the time series keys become plain tuples that are sext-encoded as part of the object.  On encode this is transparent, however on decode the riak_kv_eleveldb:from_object_key has been modified to keep the key portion as a binary regardless of whether it is a standard object or a time series object.  The key tuple can be recovered at any time useing `sext:decode` and means that any Riak code or support scripts still see decoded keys as binaries.

The time series list keys code has been modified to use the normal list keys code path and just decode the keys at the end (I think this would have been possible with the old encoding if we had written a DDL-aware decoder, but there is no type information information included in the format).

A consequence of the list keys change is that deleted objects will be included in key listing until their tombstones are removed from the underlying eleveldb database.

There are still some eunit failures that need to be resolved.  Opening to give a chance for reviewers to comment while they are fixed up.

Commit:
Added support for timeseries family/series/timestamp tuples as keys.

Alternate riak_object:newts for making objects with non-binary keys.
Customized the eleveldb:to/from_object calls to encode any key as sext,
but leave the key as a sext-encoded binary on decode.

Timing measurements (mean over 100k trials)
```
                 Previous  Current
to_object_ke       31us     29us
from_object_key    82us     71us
```

Commit:
Updated first key to work with binaries and tuple-encoded keys.

With the changes to support time series the trick of using
an empty binary for the first key sorted *after* the tuple,
so switched to using a bare atom instead.

```
(dev1@127.0.0.1)7> <<>> < {a,b,c}.
false
(dev1@127.0.0.1)8> undefined < {a,b,c}.
true
```

Commit:
Changed delete to use a riak_client:get call that permits tuple-keys.

Previously was crashing before the reap when it tried to issue the
reaping get.

Commit:
Get TS list keys working with the new key format.

No longer needs the special TS list keys coverage code
as the keys can be directly recovered.  One consequence of this
is that objects are not retrieved so deleted entries will show up
until they are reaped from the vnode like regular riak KV.

If you retrieve them with C:get(..., [deletedvclock]) you will
see they are removed but have not been reaped yet.

Commit:
Removed tests for factored out ```strip_nonlk_fields/2```

Commit:
Partial revert of riak_object:new refactor to make unit tests pass.

Tests cleverly spotted initialization of updatemetadata in the
record definition I had missed.  Now passes riak_object eunit test.

JeetKunDoug commented on src/riak_kv_eleveldb_backend.erl:
```diff
@@ -464,18 +465,18 @@ range_scan(FoldIndexFun, Buffer, Opts, #state{fold_opts=_FoldOpts,
     StartK3 = riak_ql_ddl:make_key(Mod, LK, StartK2),
-    StartKey = eleveldb_ts:encode_key(StartK3),
-    StartKey2 = to_object_key(Bucket, StartKey),
+    StartK4 = list_to_tuple([Val || {_Type, Val} <- StartK3]), %% TODO: Avoid adding/removing type info
```
this `list_to_tuple([V || {_T, V} <- Key])` should really be wrapped in a function somewhere that's well named and has some edoc about what the heck it's doing and why... it's pretty cryptic trying to understand it reading through the code.

Commit:
Added EQC test for new encoder/decoder functions and removed export_all

As requested by @macintux.

Commit:
Use table_to_bucket_name instead of creating 2-tuple for tslistkeys

Other functions still directly genreate, but as this PR does not touch them
leaving alone for a cleanup PR.

jonmeredith commented:
Pushed fixes for things that came up in the walkthrough. Will address any review feedback next.

Commit:
Refactored local/partition key to object key into function

Refactored type-stripping list comprehension and list_to_tuple
into riak_kv_ts_util:encode_typeval_key

Commit:
Changed TS delete to pass {dw,all} to work around async tombstone put/reaping get race.

As exhibited by ts_basic in riak_test.

Commit:
Dialyzer cleanups.

    Unrolled from_object no longer errors, always ignores.
    Removed extra ts list keys clauses dialyzer spotted.  Checked not in 2.1 vnode code.

Commit:
Expand on reason for using {dw,all} in PB timeseries delete.

JeetKunDoug commented:
:+1:

JeetKunDoug commented:
:+1:    66f22a9

JeetKunDoug commented:
@borshop merge

jonmeredith commented:
@borshop merge

jonmeredith commented:
@borshop merge
